src/libopensc/padding.c

```diff
@@ -653,7 +653,7 @@ int sc_get_encoding_flags(sc_context_t *ctx,
 		*sflags = SC_ALGORITHM_RSA_PAD_NONE;
 		*pflags = iflags;
 
-	} else if ((caps & (SC_ALGORITHM_RSA_PAD_PKCS1 | SC_ALGORITHM_RSA_HASH_NONE)) &&
+	} else if ((caps & SC_ALGORITHM_RSA_PAD_PKCS1) && (caps & SC_ALGORITHM_RSA_HASH_NONE) &&
```

Comment:
My problem with encoding flags is not covered with this. See my log in #2718:
```
P:20188; T:12200 2023-02-23 09:27:00.238 [cardmod] padding.c:622:sc_get_encoding_flags: iFlags 0x100002, card capabilities 0x100100
```
`caps` doesn't have `SC_ALGORITHM_RSA_PAD_PKCS1` set.

I'd rather fix this in minidriver code with something like:
```
--- minidriver.c
+++ minidriver.c
@@ -4978,6 +4978,7 @@
         dwret = SCARD_E_INVALID_VALUE;
         goto err;
     }
+    opt_crypt_flags &= ~SC_ALGORITHM_RSA_PADS;
   } else {
     logprintf(pCardData, 0, "invalid private key\n");
     dwret = SCARD_E_INVALID_VALUE;
```
and exclude any padding flags there when doing ECDSA?

Reply:
Well it does.  And it looks like minidriver sets it when it should not. 

In minidriver.c (line numbers):
4692 DWORD WINAPI CardSignData(__in PCARD_DATA pCardData, __inout PCARD_SIGNING_INFO pInfo)
4699         int opt_crypt_flags;

In one of you logs I see: "CARD_PADDING_INFO_PRESENT not set:

4789                 logprintf(pCardData, 3, "CARD_PADDING_INFO_PRESENT not set\n");
4791                 opt_crypt_flags = SC_ALGORITHM_RSA_PAD_PKCS1;

This code is from cardmod when Windows only supported RSA. This looks like the problem.

4947                 r = sc_pkcs15_compute_signature(vs->p15card, pkey, opt_crypt_flags, dataToSign, dataToSignLen, pbuf, lg, NULL);

In logL sc_pkcs15_compute_signature: ECDSA using SC_ALGORITHM_ECDSA_RAW flags before 0x00000002

the "before" is the flags before they are change. which is SC_ALGORITHM_RSA_PAD_PKCS1

So my change sort of fixed it (but may break something else), but the real fix need to be in minidriver around line 4791  which should not not do any of that if not RSA.

Is that where you wanted to make the change?

Reply:
Yes, See this commit:

https://github.com/llogar/OpenSC/commit/4c31a1d4213093886859d8bff7c03dc1f212b137

Altough this clears padding flags post festum rather than prevents setting them in the first place...